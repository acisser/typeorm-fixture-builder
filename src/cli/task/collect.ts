import { sync } from 'glob';
import { relative } from 'path';
import { collect as collectFixtures } from '../../collect';
import { Task } from './task';

/**
 * Collects fixtures from bundles and stores them on the context
 * for later use.
 *
 * @param args
 */
export const collect: Task = async args => {
  const { setStatus, context, options } = args;
  const { pattern } = options;

  const bundles = sync(pattern, {
    cwd: process.cwd(),
    absolute: true,
  });

  for (const file of bundles) {
    const path = relative(process.cwd(), file);

    setStatus(`Collecting from ${path}`);

    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-explicit-any
    context.fixtures[path] = collectFixtures(require(file)) as any[];
  }
};
