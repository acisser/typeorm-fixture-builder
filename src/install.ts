/* eslint-disable @typescript-eslint/no-explicit-any */
import { Connection } from 'typeorm';
import { persist } from './persist';

/**
 * Uses the given connection to install an array of fixtures.
 * Fixtures have to be defined by using the `fixture` function.
 *
 * @see fixture
 * @param connection Connection.
 * @param fixtures Array of fixtures.
 */
export async function install<Entity>(
  connection: Connection,
  fixtures: Entity[],
): Promise<void> {
  await connection.transaction(async manager => {
    for (const fixture of fixtures) {
      await persist(manager, fixture);
    }
  });
}
