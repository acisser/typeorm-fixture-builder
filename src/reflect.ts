/* eslint-disable @typescript-eslint/no-explicit-any */
import { Resolver } from './resolve';

export const META_PERSISTED = '__PERSISTED__';
export const META_RESOLVER = '__RESOLVER__';

export function isPersisted(target: Record<string, any>): boolean {
  return Reflect.getMetadata(META_PERSISTED, target);
}

export function isFixture(target: Record<string, any>): boolean {
  return Reflect.hasMetadata(META_PERSISTED, target);
}

export function setPersisted(
  target: Record<string, any>,
  value: boolean
): void {
  Reflect.defineMetadata(META_PERSISTED, value, target);
}

export function setResolver(
  target: Record<string, any>,
  value: Resolver | undefined
): void {
  Reflect.defineMetadata(META_RESOLVER, value, target);
}

export function getResolver(target: Record<string, any>): Resolver | undefined {
  return Reflect.getMetadata(META_RESOLVER, target);
}
