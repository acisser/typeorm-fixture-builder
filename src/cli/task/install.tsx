/* eslint-disable @typescript-eslint/no-explicit-any */
import { Newline } from 'ink';
import React from 'react';
import { EntityMetadata } from 'typeorm';
import { install as installFixtures } from '../../install';
import { Fixture, Task } from './task';

/**
 * Pretty prints a labeled entity value using JSON.stringify
 *
 * @param property Property
 * @param content Content
 */
function value(property: string, content: any) {
  return [property, JSON.stringify(content)].join(': ');
}

/**
 * Pretty prints a fixture for console output.
 *
 * @param metadata Entity metadata
 * @param entity Entity instance.
 */
function pretty(metadata: EntityMetadata, entity: Fixture): string {
  const values = Object.keys(metadata.propertiesMap).reduce<string[]>(
    (values, property) => {
      if (entity[property] === undefined) {
        return values;
      }

      if (Array.isArray(entity[property])) {
        return [
          ...values,
          value(
            property,
            entity[property].map((value: any) =>
              value?.id !== undefined ? value.id : value,
            ),
          ),
        ];
      }

      if (entity[property]?.id !== undefined) {
        return [...values, value(property, entity[property].id)];
      }

      return [...values, value(property, entity[property])];
    },
    [],
  );

  return `${entity.constructor.name}(${values.join(', ')})`;
}

/**
 * Installs previoulsy collected fixtures.
 *
 * @param args
 */
export const install: Task = async args => {
  const { setStatus, pushInfo, context } = args;
  const { connection } = context;

  setStatus('');

  if (connection) {
    for (const [path, fixtures] of Object.entries(context.fixtures)) {
      pushInfo({
        underline: true,
        text: (
          <>
            <Newline />
            Installing {fixtures.length} fixtures from {path}:
          </>
        ),
      });

      await installFixtures(connection, fixtures, (entity, skipped) => {
        const prefix = skipped ? 'ℹ [SKIPPED]' : '›';
        const content = pretty(
          connection.getMetadata(entity.constructor),
          entity,
        );

        const text = [' ', prefix, content].join(' ');
        const color = skipped ? 'blue' : 'green';

        pushInfo({
          text,
          color,
        });
      });
    }

    setStatus(false);
  }

  process.exit(0);
};
