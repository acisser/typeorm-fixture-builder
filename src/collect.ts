/* eslint-disable @typescript-eslint/no-explicit-any */
import { isFixture } from './reflect';

/**
 * Collects fixtures from an arbitrary object or array structure.
 * Any value inside that structure has to be either an array of fixtures or an object which
 * values lead to a fixture.
 *
 * @param value Object or Array to collect from.
 */
export function collect<Bundle, Entity>(value: Bundle): Entity[] {
  if (value && Array.isArray(value)) {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return collectArray(value);
  }

  if (typeof value === 'object') {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return collectObject(value);
  }

  throw new Error(`Invalid fixture definition.`);
}

/**
 * Collects fixtures from an array type.
 *
 * @param value Array.
 */
function collectArray(value: any[]): any[] {
  return value.reduce<any[]>(
    (fixtures, element) => [...fixtures, ...collect(element)],
    [],
  );
}

/**
 * Collects fixtures from an object type.
 *
 * @param value Object.
 */
function collectObject(value: any): any[] {
  if (isFixture(value)) {
    return [value];
  } else {
    return Object.values(value).reduce<any[]>(
      (fixtures, element) => [...fixtures, ...collect(element)],
      [],
    );
  }
}
