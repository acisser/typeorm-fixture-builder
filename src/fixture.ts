import { DeepPartial } from "typeorm";
import { setPersisted, setResolver } from "./reflect";
import { Resolver } from "./resolve";

export function fixture<Entity extends {}>(
  entity: new () => Entity,
  data: DeepPartial<Entity>,
  resolver?: Resolver<Entity>
): Entity {
  const instance = new entity();

  setPersisted(instance, false);
  setResolver(instance, resolver);

  return Object.assign(instance, data);
}
