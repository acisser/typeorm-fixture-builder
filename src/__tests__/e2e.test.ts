import { createConnection, Connection } from "typeorm";
import { Group } from "./entities/group";
import { User } from "./entities/user";
import { install } from "../install";
import { collect } from "../collect";
import { Profile } from "./entities/profile";
import { isPersisted } from "../reflect";
import { fixture } from "../fixture";
import { Picture } from "./entities/picture";
import { dirname, resolve } from "path";
import { exec } from "child_process";

let connection: Connection;
const database = resolve(__dirname, "..", "..", "test.db");

beforeEach(async () => {
  connection = await createConnection({
    type: "sqlite",
    database,
    entities: [Group, User, Profile, Picture],
    dropSchema: true,
    synchronize: true
  });
});

afterEach(async () => {
  if (connection && connection.isConnected) {
    await connection.close();
  }
});

describe("install", () => {
  it.each`
    scenario
    ${"./scenarios/simple/simple.bundle.ts"}
    ${"./scenarios/complex/complex.bundle.ts"}
    ${"./scenarios/imports/imports.bundle.ts"}
  `("should successfully complete scenario $scenario", async ({ scenario }) => {
    const fixtures = collect(require(scenario)) as any[];
    const fixturesByType = fixtures.reduce<{ [key: string]: any[] }>(
      (grouped, fixture) => ({
        ...grouped,
        [fixture.constructor.name]: [
          ...(grouped[fixture.constructor.name] || []),
          fixture
        ]
      }),
      {}
    );

    await install(connection, fixtures);

    for (const [group, fixtures] of Object.entries(fixturesByType)) {
      for (const fixture of fixtures) {
        expect(fixture.id).toBeDefined();
        expect(isPersisted(fixture)).toEqual(true);
      }

      expect(await connection.getRepository(group).count()).toEqual(
        fixtures.length
      );
    }
  });

  it("should fail on invalid bundle", () => {
    expect(() =>
      collect(require("./scenarios/invalid/invalid.bundle"))
    ).toThrow("Invalid fixture definition.");
  });

  describe("resolve", () => {
    it("should use and merge resolved entity", async () => {
      const resolver = jest.fn((repository, { firstName }) =>
        repository
          .createQueryBuilder("user")
          .where("user.firstName = :firstName", { firstName })
      );

      const user1 = fixture(
        User,
        { firstName: "Foo", lastName: "Bar" },
        resolver
      );
      const user2 = fixture(
        User,
        { firstName: "Foo", lastName: "Baz" },
        resolver
      );

      await install(connection, [user1, user2]);

      expect(resolver).toHaveBeenCalledTimes(2);
      expect(await connection.getRepository(User).count()).toEqual(1);
      expect(await connection.getRepository(User).findOne()).toEqual({
        id: 1,
        firstName: "Foo",
        lastName: "Baz"
      });
    });

    it("should use fixture when no entity is resolved", async () => {
      const resolver = jest.fn(repository =>
        repository
          .createQueryBuilder("user")
          .where("user.firstName = :firstName", { firstName: "Baz" })
      );

      const user1 = fixture(
        User,
        { firstName: "Foo", lastName: "Bar" },
        resolver
      );

      const user2 = fixture(
        User,
        { firstName: "Foo", lastName: "Baz" },
        resolver
      );

      await install(connection, [user1, user2]);

      expect(resolver).toHaveBeenCalledTimes(2);
      expect(await connection.getRepository(User).count()).toEqual(2);
      expect(await connection.getRepository(User).find()).toEqual([
        {
          id: 1,
          firstName: "Foo",
          lastName: "Bar"
        },
        {
          id: 2,
          firstName: "Foo",
          lastName: "Baz"
        }
      ]);
    });
  });
});

describe("cli", () => {
  const executable = resolve(__dirname, "..", "..", "bin", "./cli.js");

  beforeAll(done => {
    exec("yarn compile:test", { cwd: dirname(dirname(__dirname)) }, () =>
      done()
    );
  }, 20000);

  it("should fail with unknown options", async () => {
    expect((await run(".", ["-p", "parameter"])).code).toEqual(1);
  });

  it("should display correct help texts", async () => {
    expect((await run(".", ["--help"])).stdout).toMatchSnapshot();
    expect((await run(".", ["install", "--help"])).stdout).toMatchSnapshot();
  });

  it("should successfully install a bundle", async () => {
    const scenario = resolve(__dirname, "./scenarios/simple/simple.bundle.ts");
    const fixtures = collect(require(scenario)) as any[];
    const fixturesByType = fixtures.reduce<{ [key: string]: any[] }>(
      (grouped, fixture) => ({
        ...grouped,
        [fixture.constructor.name]: [
          ...(grouped[fixture.constructor.name] || []),
          fixture
        ]
      }),
      {}
    );

    const { code, stderr } = await run(".", ["install", "-r", scenario]);

    expect(code).toEqual(0);
    expect(stderr).toMatchSnapshot();

    for (const [group, fixtures] of Object.entries(fixturesByType)) {
      expect(await connection.getRepository(group).count()).toEqual(
        fixtures.length
      );
    }

    await connection.close();
  }, 20000);

  it("should fail on an invalid bundle", async () => {
    const scenario = resolve(
      __dirname,
      "./scenarios/invalid/invalid.bundle.ts"
    );
    const { code, stderr } = await run(".", ["install", "-r", scenario]);

    expect(code).toEqual(1);
    expect(stderr).toMatchSnapshot();
  }, 20000);

  interface Result {
    code?: number;
    error: Error | null;
    stdout: string;
    stderr: string;
  }

  function run(cwd: string, args: string[]): Promise<Result> {
    return new Promise(resolve => {
      exec(
        `node ${executable} ${args.join(" ")}`,
        { cwd },
        (error, stdout, stderr) => {
          resolve({
            code: error && error.code ? error.code : 0,
            error,
            stdout,
            stderr
          });
        }
      );
    });
  }
});
