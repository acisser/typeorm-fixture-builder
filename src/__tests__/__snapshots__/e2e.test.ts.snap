// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should display correct help texts 1`] = `
"Usage: cli [options] [command]

Options:
  -h, --help                   output usage information

Commands:
  install [options] [pattern]  Load fixtures into database. Pattern is
                               optional and can be a glob pattern. [default:
                               \\"fixtures/**/*.bundle.ts\\"
"
`;

exports[`cli should display correct help texts 2`] = `
"Usage: cli install [options] [pattern]

Load fixtures into database. Pattern is optional and can be a glob pattern. [default: \\"fixtures/**/*.bundle.ts\\"

Options:
  -r, --reset-database  Drops and synchronizes the database before loading
                        fixtures
  -c, --connection      Name of connection to use. Check the typeorm
                        documentation for further information. [default:
                        \\"default\\"]
  -m, --use-migrations  Execute migrations instead synchronization after
                        dropping the database
  -h, --help            output usage information
"
`;

exports[`cli should fail on an invalid bundle 1`] = `
"- Collecting fixtures from bundles
[31mâœ–[39m Failed to load bundle file at: src/__tests__/scenarios/invalid/invalid.bundle.ts: Invalid fixture definition.
"
`;

exports[`cli should successfully install a bundle 1`] = `
"- Collecting fixtures from bundles
[32mâœ”[39m Found 1 fixture bundles with a total of 2 fixtures
- Connecting to database
[32mâœ”[39m Connected
- Synchronizing database
[32mâœ”[39m Database reset complete
- src/__tests__/scenarios/simple/simple.bundle.ts
[32mâœ”[39m src/__tests__/scenarios/simple/simple.bundle.ts
[32mâœ”[39m Done
"
`;
